cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TxLookupProvider.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.etl.common;
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.api.TxRunnable;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.etl.api.Lookup;
+import co.cask.tephra.TransactionFailureException;
+import com.google.common.base.Function;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * Implementation of {@link Lookup} that uses {@link Transactional}.
+ * This should be used when executing lookup functions outside a transaction.
+ */
+public class TxLookupProvider extends AbstractLookupProvider {
+
+  private final Transactional tx;
+
+  public TxLookupProvider(Transactional tx) {
+    this.tx = tx;
+  }
+
+  @Override
+  public <T> Lookup<T> provide(final String table, final Map<String, String> arguments) {
+    //noinspection unchecked
+    return new Lookup<T>() {
+
+      @Override
+      public T lookup(final String key) {
+        final AtomicReference<T> result = new AtomicReference<>();
+        executeLookup(table, arguments, new Function<Lookup<T>, Void>() {
```

Comment:
You can have `executeLookup` returns T, which would make the code looks cleaner.

```java
public T lookup(final String key) {
  return executeLookup(table, arguments, new Function<Lookup<T>, T>() {
    public T apply(Lookup<T> lookup) {
      return lookup.lookup(key);
    }
  });
}
```